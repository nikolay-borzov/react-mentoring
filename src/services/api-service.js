import axios from 'axios'

const apiService = {
  init() {
    axios.defaults.baseURL = API_URL
  }
}

export default apiService
