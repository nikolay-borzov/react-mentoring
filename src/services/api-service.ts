import axios from 'axios'

const apiService = {
  init() {
    // Prevent multiple calls (happens during dev-server recompile)
    if (axios.initialized) {
      return
    }

    axios.defaults.baseURL = API_URL
    // Extract data value if data property is specified
    axios.initialized = true
  }
}

export default apiService
