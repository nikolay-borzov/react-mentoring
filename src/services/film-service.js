import axios from 'axios'

export const filmService = {
  /**
   * @param {QueryParams} queryParams
   */
  getFilms(queryParams) {
    return axios
      .get('/movies', {
        params: queryParams.getParams()
      })
      .then(({ data }) => data)
  }
}
