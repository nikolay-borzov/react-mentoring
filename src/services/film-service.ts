import axios, { AxiosResponse } from 'axios'

import { Film } from '../entities/film'

export type SearchBy = 'title' | 'genres'
export type SortBy = 'release_date' | 'vote_average'
export type SortOrder = 'desc' | 'asc'

export interface GetFilmsParams {
  /** Search value */
  search?: string
  /** Type of search (title or genres) */
  searchBy?: SearchBy
  sortBy?: SortBy
  /** Value to define sort direction - 'desc' or 'asc' */
  sortOrder?: SortOrder
  /** Array to filter by genres */
  filter?: string[]
  /** Limit amount of items in result array for pagination. `10` by default */
  limit?: number
  /** Offset in result array for pagination */
  offset?: number
}

export interface DataResponse<TData> {
  data: TData[]
  total: number
  offset: number
  limit: number
}

function extractData<T>(response: AxiosResponse<T>) {
  return response.data
}

// TODO: Replace with redux-axios-middleware?
const filmService = {
  getFilms(params: GetFilmsParams) {
    return axios
      .get<DataResponse<Film>>('/movies', {
        params
      })
      .then(extractData)
  },

  getFilm(id: number) {
    return axios.get<Film>(`/movies/${id}`).then(extractData)
  }
}

export default filmService
