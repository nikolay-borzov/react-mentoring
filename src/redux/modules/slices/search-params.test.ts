import createSlice from './search-params'
import { GetFilmsParams } from '../../../services/film-service'

const initialState: GetFilmsParams = {
  search: 'Gemini',
  searchBy: 'title',
  sortBy: 'vote_average',
  sortOrder: 'desc',
  limit: 15,
  filter: []
}

const parentState = { searchParams: initialState }

const slice = createSlice({
  id: 'SEARCH_PARAMS',
  rootSelector: (state: any) => state.searchParams,
  initialState
})

describe('SearchParams state slice', () => {
  describe('selectors', () => {
    const state = parentState
    const selectors = slice.selectors

    it(`returns 'search' param`, () => {
      expect(selectors.search(state)).toBe(state.searchParams.search)
    })

    it(`returns 'searchBy' param`, () => {
      expect(selectors.searchBy(state)).toBe(state.searchParams.searchBy)
    })

    it(`returns 'sortBy' param`, () => {
      expect(selectors.sortBy(state)).toBe(state.searchParams.sortBy)
    })

    it(`returns 'filter' param`, () => {
      expect(selectors.filter(state)).toBe(state.searchParams.filter)
    })

    it(`returns search params`, () => {
      expect(selectors.params(state)).toEqual(state.searchParams)
    })
  })

  describe('actions', () => {
    const actionTypes = slice.actionTypes
    const setParams = slice.actionCreators.setParams

    it('creates an action to set search params', () => {
      const payload: GetFilmsParams = {
        search: 'Gemini',
        searchBy: 'title'
      }
      const expectedAction = {
        type: actionTypes.SET_PARAMS,
        payload
      }

      expect(setParams(payload)).toEqual(expectedAction)
    })

    it(`accepts only allowed params`, () => {
      const expectedPayload: GetFilmsParams = {
        search: 'Gemini',
        searchBy: 'title',
        sortBy: 'release_date',
        sortOrder: 'asc',
        limit: 15
      }
      const expectedAction = {
        type: actionTypes.SET_PARAMS,
        payload: expectedPayload
      }
      const payload = {
        ...expectedPayload,
        someParameter: 'someValue'
      }

      expect(setParams(payload)).toEqual(expectedAction)
    })
  })

  describe('reducer', () => {
    const reducer = slice.reducer
    const setParams = slice.actionCreators.setParams

    it('returns initial state', () => {
      expect(
        reducer(undefined, { type: 'UNKNOWN', payload: {} })
      ).toMatchSnapshot()
    })

    it('sets search params', () => {
      const state: GetFilmsParams = { search: '', searchBy: 'title', limit: 20 }
      const action = setParams({
        search: 'Horror',
        searchBy: 'genres'
      })

      expect(reducer(state, action)).toMatchSnapshot()
    })
  })
})
