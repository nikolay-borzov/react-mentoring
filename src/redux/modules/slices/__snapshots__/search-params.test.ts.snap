// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SearchParams state slice reducer returns initial state 1`] = `
Object {
  "filter": Array [],
  "limit": 15,
  "search": "Gemini",
  "searchBy": "title",
  "sortBy": "vote_average",
  "sortOrder": "desc",
}
`;

exports[`SearchParams state slice reducer sets search params 1`] = `
Object {
  "limit": 20,
  "search": "Horror",
  "searchBy": "genres",
}
`;
